### ISSUE TITLE
Decorator factory pattern

### ISSUE THREAD

**Issue by JeppeKlitgaard** [google/jax#10061]
I couldn't find a pre-existing issue for this, but I figure it might be useful to add support for the decorator factory pattern for the most common transformations (`grad, jit, vmap, pmap`, etc).

Currently people wanting to decorate functions have to either do:
```python
from functools import partial
from jax import jit

@partial(jit, param1=..., param2=...)
def foo(bar):
	...
```

Or 
```python
from jax import jit

def foo(bar):
	...

foo = jit(foo, param1=..., param2=...)
```

This seems needlessly verbose, but I am wondering if perhaps there is some rationale behind this that I might not be aware of. Being used to `numba`'s `jit` decorator, I would expect the decorator factory pattern to work.

I think it would be super neat if the following pattern was supported:
```python
from jax import jit

@jit(param1=..., param2=...)
def foo(bar):
	...
```

**Comment by JeppeKlitgaard** [comment]
I just found https://github.com/google/jax/issues/1936, which seems to be similar. 

I'm wondering if it might be worth revisiting this now that Py2 is dead and inspect might have seen improvements, so I'll leave it open for now.

**Comment by hawkinsp** [comment]
#1936 and #1159 are about something a bit different, which is static keyword arguments to the `jit`-decorated function.

We have discussed the issue you are raising (requiring `partial` as a decorator to pass arguments to `jit` itself) in the past, and it is a thing about which reasonable people disagree. If I were trying to describe the two points of view they would be:

a) "I just want to pass extra arguments to `jit` without writing `partial` (your request) by having `jit` act differently based on whether it is passed a function or keyword arguments." To some, this feels more succinct and "Pythonic".

b) "The semantics of `jit` are simpler if `jit` always acts the same regardless of the type of its arguments." While the idiom of using `partial` as a decorator is somewhat unusual, it's quite understandable once you get used to it and not *that* verbose. This approach does not require `jit` to act differently based on its arguments, and having `jit` with simpler semantics is better in the end.

In our previous debates, we've come down on the side of (b). What do you think?

**Comment by JeppeKlitgaard** [comment]
I can definitely see the viewpoint of (b), though it isn't something I have stumbled upon in other Python libraries, which is perhaps why it feels a little odd and caught me by surprise. While I also find the 'second order' decorator functions a bit strange to write (somehow writing second order decorators feels very unpythonic, but using them very pythonic).

The semantics of (a) shouldn't be too messy though – wouldn't it just be an outer wrapper that checks if first argument is callable and then calls an inner function in the appropriate manner? 

I don't have terribly strong feelings about this, but I think it would be very helpful if an example using `functools.partial` to pages like https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html. My knee-jerk reaction when I first saw the `partial` approach was something on the order of "that seems a little strange, is that really the 'conventional' way to do it?"

**Comment by jakevdp** [comment]
Judging by the thumbs up on this comment, it seems others agree with you on the need for docs: https://github.com/google/jax/issues/184#issuecomment-592753357

I'll ork on adding it.

**Comment by jakevdp** [comment]
#10074 adds some documentation of the `@partial(jit, ...)` pattern in a couple places.

**Comment by YouJiacheng** [comment]
maybe we can do some currying.
```python
import jax
def jit_currying(f=None, **kwargs):
    if f is None:
        return lambda fun: jax.jit(fun, **kwargs):
    return jax.jit(f, **kwargs)
```

**Comment by jakevdp** [comment]
```python
jit_currying = lambda *a, **k: jit(*a, **k) if a else partial(jit, *k)
```
⛳🏌️‍♂️

**Comment by msegado** [comment]
Came here to suggest this and found @JeppeKlitgaard had beat me to it... don't want to bikeshed though so please close this if you prefer to focus on other things now that the current behavior is documented!

That said, one argument in favor of supporting the decorator-or-factory pattern is that the Python standard library does so in several places:

* https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass (<= particularly prominent)
* https://docs.python.org/3/library/functools.html#functools.lru_cache
* https://docs.python.org/3/library/xmlrpc.server.html#xmlrpc.server.SimpleXMLRPCServer.register_function

I'd argue that the convenience is worth it given that the jax team itself makes use of both `@dataclass` and `@dataclass(...)` in their source code 🙂

**Comment by dibgerge** [comment]
I got here because I am new to Jax and reading the docs, I stumbled upon this syntax `partial(jit, ...)`, and it tripped me up, because I have never seen this approach before. I thought there is some fundamental limitation to take this approach and I wanted to understand it. I see this thread is a bit old, although there have been some more recent comments in the linked issues. 

What I got from this thread is that it was more of a design decision than any fundamental limitation required by jax's decorators. 

I would argue that the two points raised by @hawkinsp are not in contention, but the actual issue is "do we want behavior to change when the functions are used as decorators or called directly". This is because we need `partial` when used in decorator form due to lack of support of unspecified `function` to decorate.

Take these two implementations for example:

```python
def my_jit1(*,
  in_shardings = sharding_impls.UNSPECIFIED,
  out_shardings = sharding_impls.UNSPECIFIED,
  static_argnums = None,
  static_argnames = None,
  donate_argnums= None,
  donate_argnames = None,
  keep_unused: bool = False,
  device = None,
  backend: str | None = None,
  inline: bool = False,
  abstracted_axes = None,
  compiler_options = None,
):
    def decorator(fun):
        return pjit.make_jit(fun, in_shardings=in_shardings, out_shardings=out_shardings, static_argnums=static_argnums, static_argnames=static_argnames, donate_argnums=donate_argnums, donate_argnames=donate_argnames, keep_unused=keep_unused, device=device, backend=backend, inline=inline, abstracted_axes=abstracted_axes, compiler_options=compiler_options, use_resource_env=False)
      
   return decorator

def my_jit2(fun= None, /, *, ... )
  """ `...` are same kwargs as in my_jit1 """
    def decorator(fun):
        return pjit.make_jit(fun, ..., use_resource_env=False)
      
    if fun is None:
      return decorator

    return decorator(fun)
```


## Usage as a decorator

This should work exactly like the current implementation:

```python

# current implementation
@jax.jit
def some_function(...):
 ....

# current implementation with arguments
@partial(jax.jit, static_argnums=0):
def some_function(...):
  ...

# proposed implementation (no parenthesis) **does not work** for my_jit1, but **works** for my_jit2
@my_jit2
def some_fuction(...):
  ...

# proposed implementation (with parenthesis) works also for my_jit1
@my_jit2()
def some_function(...):
  ...

# proposed implementation with arguments works also same for my_jit1
@my_jit2(static_argnums=0)
def some_function(...):
  ...

type(some_function)

# jaxlib.xla_extension.PjitFunction
# (for all cases)
```

So the new implementation works without the need of `partial`, but we have to decide the convenience to make it work by adding the parenthesis to the decorator with no arguments (i.e. using code of `my_jit2`) or not. This is because `my_jit2` implementation has implications on direct function call as seen below.

## Direct function call 

The following two usages would be allowed for both `my_jit1`, where arguments are given first, and then a new set of parenthesis for the function. Can't provide the function in the first call:

```
my_jit1()(some_function)
my_jit1(static_argnums=0)(some_function)

# not allowed
my_jit1(some_function)
```

The upside is that `my_jit1` will always have same behavior.

`my_jit2` is simpler in this case, because we only have two use on set of parenthesis, but depending if you give a function or not in the first argument the behavior will change. The following are all allowed

```
my_jit2()(some_function)
my_jit2(static_argnums=0)(some_function)

my_jit2(some_function)
my_jit2(some_function, static_argnums=0)
```

## Summary

To summarize, my opinion is that it is better chose between the two implementations of `my_jit1` and `my_jit2` than having to rely on `partial`. This could be based on what is more commonly used: the decorator form (use `my_jit2` has no side effects)  vs the direct call form (use of `my_jit1` has no side effects).  

Also, this is appears to be what the python developers do, as @msegado  has indicated, where `dataclass` would actually return different types depending on when you give it a class or not, and uses an implementation similar to  `my_jit2`.

**Comment by yashk2810** [comment]
@jakevdp has a PR out to implement the decoratory factory for `jax.jit` but it's blocked by some pytype failures. Once those are resolved, we should be able to submit that PR.

**Comment by jakevdp** [comment]
For JIT, this is done in #32861, and should be available starting in JAX v0.8.1.

We've discussed elsewhere that the decorator factory pattern makes less sense for `vmap` and `grad`, and we don't have plans for it there. I think we can close this!

**Comment by cool-RR** [comment]
I'm happy to see this, thank you for implementing!
